### PR TITLE
Spectrum1D output to text file

### PR DESCRIPTION
**Pull request by rebeccafair**
Addresses #190 

A few things to note:
* Note that if the spectrum's `x_data` contains bin edges they need to be converted to bin centres before writing, as we need the same number of x and y data points. This means that the written spectrum will not be exactly the same in these cases.
* I decided since we can write to these files we should probably be able to read from them too, so have added that in. This means we may have to be slightly stricter about the header if we want to be able to consistently read it again. For this I thought it made sense to have the metadata as a JSON string. Currently the header lines can be in any order (not sure if this is a good thing..), and if information is not found in the header defaults will be set (dimensionless units and empty metadata), so we can just read in columns of data with no extra information if we want to. A good example of what the full header looks like is in `tests_and_analysis/test/data/spectrum1dcollection/quartz_666_coh_pdos.text`. Any thoughts on this let me know.
* I also couldn't decide on the f.p output format (the 15+ s.f. default seemed excessive) so have allowed the user to pass in a format string straight to `np.savetxt`.  Maybe this is overkill as we don't do any f.p. formatting on the json files currently.

### CODE REVIEW

**Review by mducle — APPROVED** [review]
Looks good to me, and works well.

The only comment I have is to maybe think about adding a check for a `# Generated by Euphonic` header in the `from_text_file` method and if not possibly emit a warning... but this might not be necessary.

**Comment by ajjackson** [comment]
I don't see how we can simultaneously:
- support import from text files
- allow arbitrary number formatting in said text files

**Review by ajjackson — COMMENTED** [review]
Implementation looks fine.

I'm not sure that it's a good idea to read the text files, except for purposes of testing. We already have a machine-readable JSON output which is easy to maintain. By supporting read from this text output we discourage users from making use of JSON outputs, and lock ourselves into supporting/maintaining a more complicated and fragile text parser.

**Comment by mducle** [comment]
> I don't see how we can simultaneously:
> 
>     * support import from text files
> 
>     * allow arbitrary number formatting in said text files

Yeah, good point. I guess there could be a check when writing which would warn users that this file cannot be read back by Euphonic, e.g.

```
from tempfile import TemporaryFile
import warnings
with TemporaryFile() as tf:
    dos_sib.to_text_file(tf, fmt=['%g', 'a%s'])
    tf.seek(0)
    try:
        dosreload = np.loadtxt(tf)
    except ValueError:
        warnings.warn('Warning: Saved files with this format cannot be re-read by Euphonic')
```

But you're right that it would probably be easier if we don't have a reader, or at least don't publicise the reader (maybe make it dunder method?)

**Comment by ajjackson** [comment]
The code could live in test utilities as a pure function returning an instantiated Spectrum1DCollection.

**Comment by rebeccafair** [comment]
Ok fair points, I've moved the text file reading into test utilities, and made it less flexible so requires a particular header order. I guess the fact that I spent so long umming and ahing about the header reading should probably have been a red flag.

File: euphonic/spectra.py [hunk 1]
```diff
@@ -403,6 +422,33 @@ def from_dict(cls: Type[T], d: Dict[str, Any]) -> T:
...
+        # y_data in (millibarn / millielectron_volt)
+        # Common metadata: {"weighting": "coherent", "species": "O"}
+
+        Parameters
+        ----------
+        filename
+            Name of the text file to read from
+        """
+        spec = Spectrum1DCollection.from_text_file(filename)
+        return spec[0]
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/spectra.py]: `from_text_file` has now been removed

File: euphonic/spectra.py [hunk 2]
```diff
@@ -733,6 +779,41 @@ def to_dict(self) -> Dict[str, Any]:
...
+        filename
+            Name of the text file to write to
+        fmt
+            A format specifier or sequence of specifiers (one for each
+            column), to be passed to numpy.savetxt
+        """
+        common_metadata = copy.deepcopy(self.metadata)
+        line_data = common_metadata.pop('line_data',
+                                        [{} for i in self.y_data])
+        header = []
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/spectra.py]: No idea, I think it's just the first thing that came to mind. Now fixed.

File: tests_and_analysis/test/utils.py [hunk 3]
```diff
@@ -304,7 +304,7 @@ def get_spectrum_from_text(text_filename, is_collection=True):
         metadata.update(json.loads(line[line.index('{'):]))
         fp.readline()  # Skip 'Column 1: x_data' line
         for idx in range(len(y_data[0])):
-            line = fp.readline()  # 'y_data in' line
+            line = fp.readline()  # 'Column i: y_data[i + 1]...' line
```
Comment (ajjackson) [inline tests_and_analysis/test/utils.py:307]: Isn't it "Column i+2: y_data[i]"?

**Review by ajjackson — APPROVED** [review]
See query above about indexing in a comment.

Otherwise seems good!

File: tests_and_analysis/test/utils.py (hunk 3 above)
Comment (rebeccafair) [reply tests_and_analysis/test/utils.py]: Yes.

**Pull request merged**